### PR TITLE
`ultralytics 8.3.207` Apple MLProgram NMS models 80-class pad workaround

### LINKED ISSUE #19857: The model converted to coreml format always shows confidence 1.0
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

_No response_

### Bug

The model converted to coreml format always shows confidence 1.0   

### Environment

The model converted to coreml format always shows confidence 1.0

### Minimal Reproducible Example

The model converted to coreml format always shows confidence 1.0

### Additional

_No response_

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

### LINKED ISSUE #20061: NMS Export Results in Spurious Detections with Custom YoloV8m model.
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

This issue has been happening to me since the export moved to generating via CoreMLTools 8.x.  It is also a followup to this [ticket](https://github.com/ultralytics/ultralytics/issues/19953).

I am exporting a model that I trained via the Ultralytics HUB and bring your own runner. When I download the model to the HUB app it works as expected. When I export the model, whether via the HUB interface or via the python with the Ultralytics package, exporting with NMS enabled results in spurious detections. This is a screen capture from loading the model into Xcode and previewing it.

Image

My hunch is that there is something wrong with the cast operations being inserted right before the NMS operation. 

Image

I can provide the source PyTorch model or the Hub model ID if that would help aid debugging. 

For reference, this is my export script:
```
from ultralytics import YOLO, checks, hub
checks()

hub.login('<id>')

model = YOLO('https://hub.ultralytics.com/models/<modelid>')
model.export(format="coreml",nms=True,half=False)
```
I have also tried exporting by modifying the model loading line to `model = YOLO('https://hub.ultralytics.com/models/<modelid>').float()` and that does not seem to make a difference.

### Environment

Ultralytics 8.3.103 🚀 Python-3.11.5 torch-2.3.1 CPU (Apple M1 Max)
Setup complete ✅ (10 CPUs, 64.0 GB RAM, 1605.3/1858.2 GB disk)

OS                  macOS-15.4-arm64-arm-64bit
Environment         Darwin
Python              3.11.5
Install             git
Path                /Users/davidsiebecker/Developer/ultralytics/ultralytics
RAM                 64.00 GB
Disk                1605.3/1858.2 GB
CPU                 Apple M1 Max
CPU count           10
GPU                 None
GPU count           None
CUDA                None

numpy               ✅ 1.23.5<=2.1.1,>=1.23.0
matplotlib          ✅ 3.8.4>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.13.0>=1.4.1
torch               ✅ 2.3.1>=1.8.0
torch               ✅ 2.3.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.18.1>=0.9.0
tqdm                ✅ 4.66.2>=4.64.0
psutil              ✅ 5.9.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.1>=1.1.4
seaborn             ✅ 0.12.2>=0.11.0
ultralytics-thop    ✅ 2.0.0>=2.0.0

### Minimal Reproducible Example

```
from ultralytics import YOLO, checks, hub
checks()

hub.login('<id>')

model = YOLO('https://hub.ultralytics.com/models/<modelid>')
model.export(format="coreml",nms=True,half=False)
```

### Additional

_No response_

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

### LINKED ISSUE #22309: CoreML export of YOLOv11 model produces incorrect or missing bounding boxes, while TFLite works as expected
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

I trained a YOLOv11 model using the same dataset and training parameters. When exporting the trained model to both TFLite and CoreML formats, I observed the following issue:

The TFLite model returns correct detection results.
The CoreML model either returns no detections or produces bounding boxes that are missing or located outside the image boundaries.

I've attached the example notebook that I'm using. And two images of the output as a reference.

Image
Image

[Train_YOLO_Models.ipynb](https://github.com/user-attachments/files/22751535/Train_YOLO_Models.ipynb)

### Environment

```
Ultralytics 8.3.206 🚀 Python-3.11.11 torch-2.8.0 CPU (Apple M4 Pro)
Setup complete ✅ (14 CPUs, 48.0 GB RAM, 758.9/926.4 GB disk)

OS                     macOS-26.0.1-arm64-arm-64bit
Environment            Darwin
Python                 3.11.11
Install                pip
Path                   /Users/jandrop/Developer/AI/Yolo/.venv/lib/python3.11/site-packages/ultralytics
RAM                    48.00 GB
Disk                   758.9/926.4 GB
CPU                    Apple M4 Pro
CPU count              14
GPU                    None
GPU count              None
CUDA                   None

numpy                  ✅ 2.1.3>=1.23.0
matplotlib             ✅ 3.10.6>=3.3.0
opencv-python          ✅ 4.12.0.88>=4.6.0
pillow                 ✅ 11.3.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.16.2>=1.4.1
torch                  ✅ 2.8.0>=1.8.0
torch                  ✅ 2.8.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.23.0>=0.9.0
psutil                 ✅ 7.1.0
polars                 ✅ 1.34.0
ultralytics-thop       ✅ 2.0.17>=2.0.0
```

### Minimal Reproducible Example

The notebook with the reproducible code is attached

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes #22309 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Fixes CoreML export NMS issues for detection models on iOS, especially with non-80 class counts, and simplifies weights handling during export. 🚀

### 📊 Key Changes
- Version bump to 8.3.207.
- CoreML export:
  - IOSDetectModel now receives an `mlprogram` flag to handle MLProgram-specific behavior.
  - Pads class scores to the next multiple of 80 for MLProgram exports when NMS is enabled, avoiding a known CoreML NMS issue.
  - Automatically adjusts class name mapping if the exported model’s class count and provided names differ.
  - Simplifies `weights_dir` handling by using `ct_model.weights_dir` directly, removing manual save-and-path logic.

### 🎯 Purpose & Impact
- Resolves the MLProgram NMS bug for detection models with arbitrary class counts, improving reliability on iOS devices. 🛠️
- Ensures correct label handling even when class name lists don’t match the model’s output size, preventing runtime errors. ✅
- Streamlines CoreML export pipeline, reducing edge-case failures and making on-device deployment smoother. 📱

Tip: For detection models, use NMS in CoreML export like this:
```bash
yolo export model=yolo11n.pt format=coreml nms=True
```

Reference: See the MLProgram NMS bug details in the related issue: MLProgram NMS bug.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline the review process—an Ultralytics engineer will also assist you here shortly 🙂

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Additional notes for this PR:
- Great call-out on CoreML NMS behavior and class-count padding. Please ensure tests cover:
  - Export with `nms=True` for class counts not divisible by 80, confirming no NMS failures
  - Export with `nms=False` to verify no regression in default paths
  - Metadata handling when `len(names) != nc` to validate numeric backfilling
- If applicable, a brief documentation note on the CoreML export behavior (class padding to multiples of 80) would be helpful for users.

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀🍏

**Comment by Y-T-G** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch on this PR branch `coreml-nms-custom` at 2025-10-08 02:10:57 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/18331593557): `ci.yml` run 12831
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/18331593691): `docker.yml` run 1235

**Comment by UltralyticsAssistant** [comment]
Huge thanks for the merge, @Y-T-G — and kudos to @glenn-jocher for the support! “If I have seen further, it is by standing on the shoulders of giants.” — Isaac Newton. This PR meaningfully elevates our iOS pipeline: robust CoreML NMS with MLProgram for arbitrary class counts, smarter class-name mapping, and cleaner weight handling, all shipping with 8.3.207. Your attention to detail directly improves on-device reliability and developer experience. 

Quick tip for detection models: enable NMS during export with `yolo export model=yolo11n.pt format=coreml nms=True`.  
Learn more in the Ultralytics CoreML export docs: [CoreML export guide](https://docs.ultralytics.com/modes/export/#coreml).

Appreciate the craftsmanship—this will help countless users deploy YOLO11 models smoothly on iOS. 🚀

**Comment by Y-T-G** [comment]
Refound this old Reddit thread where the issue originated in MacOS 15.2

https://www.reddit.com/r/Ultralytics/comments/1hahp6r/broken_coreml_models_on_macos_152/
